x-pack/plugins/ml/public/application/services/dashboard_service.ts

```diff
@@ -0,0 +1,113 @@
+import {
+  DASHBOARD_APP_URL_GENERATOR,
+  DashboardUrlGenerator,
+  SavedDashboardPanel,
+  SavedObjectDashboard,
+} from '../../../../../../src/plugins/dashboard/public';
+import { useMlKibana } from '../contexts/kibana';
+import { ViewMode } from '../../../../../../src/plugins/embeddable/public';
+
+export type DashboardService = ReturnType<typeof dashboardServiceProvider>;
+
+export function dashboardServiceProvider(
+  savedObjectClient: SavedObjectsClientContract,
+  kibanaVersion: string,
+  dashboardUrlGenerator: DashboardUrlGenerator
+) {
+  const generateId = htmlIdGenerator();
+
+  return {
+    /**
+     * Fetches dashboards
+     */
+    async fetchDashboards(query?: string) {
+      return await savedObjectClient.find<SavedObjectDashboard>({
+        type: 'dashboard',
+        perPage: 10,
+        search: query ? `${query}*` : '',
+        searchFields: ['title^3', 'description'],
+      });
+    },
+    /**
+     * Attaches embeddable panels to the dashboard
+     */
+    async attachPanels(
+      dashboardId: string,
+      dashboardAttributes: SavedObjectDashboard,
+      panelsData: Array<Pick<SavedDashboardPanel, 'title' | 'type' | 'embeddableConfig'>>
+    ) {
+      const panels = JSON.parse(dashboardAttributes.panelsJSON) as SavedDashboardPanel[];
+      const version = kibanaVersion;
+
+      for (const panelData of panelsData) {
+        const panelIndex = generateId();
+        const maxPanel =
+          panels.length > 0
+            ? panels.reduce((prev, current) =>
+                prev.gridData.y > current.gridData.y ? prev : current
+              )
+            : null;
+        panels.push({
+          panelIndex,
+          embeddableConfig: panelData.embeddableConfig as { [key: string]: any },
+          title: panelData.title,
+          type: panelData.type,
+          version,
+          gridData: {
+            h: 15,
+            i: panelIndex,
+            w: 24,
+            x: 0,
+            y: maxPanel ? maxPanel.gridData.y + maxPanel.gridData.h : 0,
+          },
+        });
+      }
+
+      await savedObjectClient.update('dashboard', dashboardId, {
+        ...dashboardAttributes,
+        panelsJSON: JSON.stringify(panels),
+      });
+    },
+    /**
+     * Generates dashboard url with edit mode
+     */
+    async getDashboardEditUrl(dashboardId: string) {
+      return await dashboardUrlGenerator.createUrl({
+        dashboardId,
+        useHash: false,
+        viewMode: ViewMode.EDIT,
+      });
+    },
+  };
+}
+
+/**
+ * Hook to use {@link DashboardService} in react components
+ */
+export function useDashboardService(): DashboardService {
+  const {
+    services: {
+      savedObjects: { client: savedObjectClient },
+      kibanaVersion,
+      share: { urlGenerators },
+    },
+  } = useMlKibana();
+  return useMemo(
+    () =>
+      dashboardServiceProvider(
+        savedObjectClient,
+        kibanaVersion,
+        urlGenerators.getUrlGenerator(DASHBOARD_APP_URL_GENERATOR)
```

Comment:
FYI, we trying to go away from this pattern with getById() in favour of explicit usage from plugin contract. 
Mostly due to better type safety when using this. 
see example pr: https://github.com/elastic/kibana/pull/64628